velox/type/Filter.cpp

```diff
@@ -458,6 +470,75 @@ bool MultiRange::testBytesRange(
   return false;
 }
 
+std::unique_ptr<Filter> MultiRange::mergeWith(const Filter* other) const {
+  switch (other->kind()) {
+    case FilterKind::kMultiRange: {
+      bool bothNullAllowed = nullAllowed_ && other->testNull();
+      // TODO: verify nanAllowed with the other filter.
+      const MultiRange* multiRangeOther = static_cast<const MultiRange*>(other);
+      std::vector<std::unique_ptr<Filter>> merged;
+      for (auto const& filter : this->filters()) {
+        for (auto const& filterOther : multiRangeOther->filters()) {
+          auto innerMerged = filter->mergeWith(filterOther.get());
+          switch (innerMerged->kind()) {
+            case FilterKind::kAlwaysFalse:
+              break;
+            case FilterKind::kIsNotNull:
+            case FilterKind::kIsNull:
+              // the merged filter is valid when it matches with top-level
+              // nullAllowed.
+              if (innerMerged->testNull() && bothNullAllowed) {
+                merged.push_back(std::move(innerMerged));
+              }
+              break;
+            default:
+              merged.push_back(std::move(innerMerged));
+          }
+        }
+      }
+
+      if (merged.empty()) {
+        return nullOrFalse(bothNullAllowed);
+      } else if (merged.size() == 1) {
+        // If both sides allow null while the merged inner filter does not allow
+        // null, then this filter should be treated as isNULL.
+        if (bothNullAllowed && !merged[0]->testNull()) {
+          merged.pop_back();
+          merged.push_back(std::make_unique<IsNull>());
+        }
+      }
+      return std::unique_ptr<Filter>(
+          new MultiRange(std::move(merged), bothNullAllowed, nanAllowed_));
+    }
+    case FilterKind::kAlwaysTrue:
+    case FilterKind::kAlwaysFalse:
+    case FilterKind::kIsNull:
+      return other->mergeWith(this);
+    case FilterKind::kIsNotNull: {
+      std::unique_ptr<Filter> clonedFilter = this->clone();
```

Comment:
can we do this only if this filter's nullAllowed is false?

Reply:
I see. So `null` AND `not null` should be null still. Only `not null` AND `not null` will be `not null`?

Sure but what to return when it is `null` AND `not null`?

Reply:
I am thinking it should just return this->clone() when `null` ADN `not null`.

Reply:
Here is how I'm thinking about this. There are 2 possibilities for a MultiRange. Either it allows a null value or not.

- MultiRange(nullAllowed = true) AND IS NULL => IS NULL
- MultiRange(nullAllowed = true) AND IS NOT NULL => MultiRange(nullAllowed = false)
- MultiRange(nullAllowed = false) AND IS NULL => ALWAYS FALSE
- MultiRange(nullAllowed = false) AND IS NOT NULL => MultiRange(nullAllowed = false)

Does this make sense?

Reply:
Excellence summary. I even added this summary as comments in code.